gh-135700: Fix instructions in __annotate__ have incorrect code positions

AndPuQing opened the pull request:
<!-- gh-issue-number: gh-135700 -->
* Issue: gh-135700
<!-- /gh-issue-number -->

```python
from dis import Bytecode, dis

src = """\
22
333
__dataclass_fields__: ClassVar
"""

code = compile(src, "<string>", "exec")
print(code.co_consts)
print(dis(src))
annotate_code = code.co_consts[1]

for bc in Bytecode(annotate_code):
    print(bc.positions, bc.opname)
```
output

```python
(0, <code object __annotate__ at 0x7b19f4095f20, file "<string>", line 1>, None)
  --           MAKE_CELL                0 (__conditional_annotations__)

   0           RESUME                   0

   1           LOAD_CONST               1 (<code object __annotate__ at 0x7b19f4096c30, file "<dis>", line 1>)
               MAKE_FUNCTION
               STORE_NAME               1 (__annotate__)
               BUILD_SET                0
               STORE_NAME               0 (__conditional_annotations__)

   2           NOP

   3           LOAD_NAME                0 (__conditional_annotations__)
               LOAD_SMALL_INT           0
               SET_ADD                  1
               POP_TOP
               LOAD_CONST               2 (None)
               RETURN_VALUE

Disassembly of <code object __annotate__ at 0x7b19f4096c30, file "<dis>", line 1>:
  1           RESUME                   0
              LOAD_FAST_BORROW         0 (format)
              LOAD_SMALL_INT           2
              COMPARE_OP             132 (>)
              POP_JUMP_IF_FALSE        3 (to L1)
              NOT_TAKEN
              LOAD_COMMON_CONSTANT     1 (NotImplementedError)
              RAISE_VARARGS            1
      L1:     BUILD_MAP                0

  3           LOAD_SMALL_INT           0
              LOAD_GLOBAL              0 (__conditional_annotations__)
              CONTAINS_OP              0 (in)
              POP_JUMP_IF_FALSE       10 (to L2)
              NOT_TAKEN
              LOAD_GLOBAL              2 (ClassVar)
              COPY                     2
              LOAD_CONST               1 ('__dataclass_fields__')

  1           STORE_SUBSCR
      L2:     RETURN_VALUE
None
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) RESUME
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) LOAD_FAST_BORROW
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) LOAD_SMALL_INT
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) COMPARE_OP
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) POP_JUMP_IF_FALSE
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) NOT_TAKEN
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) LOAD_COMMON_CONSTANT
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) RAISE_VARARGS
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) BUILD_MAP
Positions(lineno=3, end_lineno=3, col_offset=0, end_col_offset=30) LOAD_SMALL_INT
Positions(lineno=3, end_lineno=3, col_offset=0, end_col_offset=30) LOAD_GLOBAL
Positions(lineno=3, end_lineno=3, col_offset=0, end_col_offset=30) CONTAINS_OP
Positions(lineno=3, end_lineno=3, col_offset=0, end_col_offset=30) POP_JUMP_IF_FALSE
Positions(lineno=3, end_lineno=3, col_offset=0, end_col_offset=30) NOT_TAKEN
Positions(lineno=3, end_lineno=3, col_offset=22, end_col_offset=30) LOAD_GLOBAL
Positions(lineno=3, end_lineno=3, col_offset=0, end_col_offset=30) COPY
Positions(lineno=3, end_lineno=3, col_offset=0, end_col_offset=30) LOAD_CONST
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) STORE_SUBSCR
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) RETURN_VALUE
```

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

AndPuQing committed:
fix instructions in __annotate__ have incorrect code positions

sobolevn commented on Misc/NEWS.d/next/Core_and_Builtins/2025-06-22-10-47-27.gh-issue-135700.0qdtCl.rst:
```diff
@@ -0,0 +1 @@
+Fix instructions in __annotate__ have incorrect code positions
```
```suggestion
Fix instructions positions in :attr:`~object.__annotate__`.
```

AndPuQing committed:
Update Misc/NEWS.d/next/Core_and_Builtins/2025-06-22-10-47-27.gh-issue-135700.0qdtCl.rst

Co-authored-by: sobolevn <mail@sobolevn.me>

picnixz committed:
gh-135239: simpler use of mutexes in cryptographic modules (#135267)

AndPuQing committed:
add unit tests

sobolevn commented on Lib/test/test_dis.py:
```diff
@@ -431,6 +431,40 @@ def foo(a: int, b: str) -> str:
+Disassembly of __annotate_func__:
+  --           COPY_FREE_VARS           1
+
+ %3d           RESUME                   0
```
Maybe we don't need all instructions here? We only care about positions.

picnixz committed:
gh-135815: skip `netrc` security checks if `os.getuid` is missing (#135816)

picnixz committed:
gh-135532: cleanup clinic `module` directives for cryptographic modules (#135822)

AndPuQing committed:
fix test

kumaraditya303 committed:
add async generators section to asyncio internal docs  (#135674)

iritkatriel commented on Lib/test/test_dis.py:
```diff
@@ -1124,6 +1136,28 @@ def test_bug_46724(self):
         # Test that negative operargs are handled properly
         self.do_disassembly_test(bug46724, dis_bug46724)
 
+    def test_annotate_no_spurious_first_node_positions(self):
```
Can the test be written to assert the correct position (rather than assert it's not an incorrect value?)

emmatyping committed:
gh-134986: Catch PermissionError when trying to call perf in tests (#134987)

Using Ubuntu 24.04 on the Windows Subsystem for Linux, perf will raise a
`PermissionError` instead of `FileNotFoundError`. This commit modifies
the tests to catch that.

encukou committed:
gh-89488: Add warning about `Py_BuildValue("p")` needing exact `int` (GH-135610)

picnixz committed:
gh-135823: improve error message in `netrc` security checks (#135827)

AndPuQing committed:
fix test

sobolevn committed:
Bump mypy to 1.16.1 (#135720)

kumaraditya303 committed:
GH-124878: reenable `test_finalize_daemon_thread_hang` test under TSAN (#135793)

Fidget-Spinner committed:
gh-134584: Specialize POP_TOP by reference and type in JIT (GH-135761)

iritkatriel commented on Lib/test/test_dis.py:
```diff
@@ -1124,6 +1124,47 @@ def test_bug_46724(self):
         # Test that negative operargs are handled properly
         self.do_disassembly_test(bug46724, dis_bug46724)
 
+    def test_annotate_no_spurious_first_node_positions(self):
```
```suggestion
    def test_annotate_source_locations(self):
```

iritkatriel commented on Lib/test/test_dis.py:
```diff
@@ -1124,6 +1124,47 @@ def test_bug_46724(self):
         self.do_disassembly_test(bug46724, dis_bug46724)
 
+    def test_annotate_no_spurious_first_node_positions(self):
+        # Test that __annotate__ code doesn't inherit first AST node positions
```
```suggestion
        # See gh-135700
```

iritkatriel commented on Lib/test/test_dis.py:
```diff
@@ -1124,6 +1124,47 @@ def test_bug_46724(self):
+
+        for case_name, annotate_code in test_cases:
+            with self.subTest(case=case_name):
+                instructions = list(dis.Bytecode(annotate_code))
```
You can use `dis.findlinestarts(annotate_code)` instead.

iritkatriel commented on Lib/test/test_dis.py:
```diff
@@ -1124,6 +1124,47 @@ def test_bug_46724(self):
+        for case_name, annotate_code in test_cases:
+            with self.subTest(case=case_name):
+                instructions = list(dis.Bytecode(annotate_code))
+                print(instructions)
```
```suggestion
```

noamcohen97 committed:
gh-131798: Optimize `_UNARY_NEGATIVE` (GH-135223)

nascheme committed:
gh-119786: Add InternalDocs/qsbr.md. (gh-135411)

Add internal doc for the Quiescent-State Based Reclamation (QSBR) implementation.

Yzi-Li committed:
Docs: Use `arguments` to replace `args` in `argparse.rst` (GH-135510)

vpoulailleau committed:
Fix example according to PEP 750 in "What's new in 3.14" (GH-134727)

A redundant extra part was written. Added a closing tag, to match the usage in PEP 750.

freakboy3742 committed:
gh-135648: Document that `shutil.copyfileobj` doesn't flush (#135737)

Adds a note about flush/close on copyfileobj, and updates 
the Emscripten build script to follow documented advice.

AndPuQing committed:
Update Lib/test/test_dis.py

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

AndPuQing committed:
Apply suggestions from code review

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

picnixz committed:
gh-135532: use `defining_class` for copying BLAKE-2 and SHA-3 objects (#135838)

hugovk committed:
gh-123299: Update 'What's New in Python 3.14' from `3.14` branch (#135616)

AndPuQing committed:
fix test

markshannon committed:
GH-135106: Restrict trashcan to GC'ed objects (GH-135682)

vstinner committed:
gh-135494: Fix python -m test --pgo -x test_re (#135713)

Fix regrtest to support excluding tests from --pgo tests.

picnixz committed:
gh-135487: fix `reprlib.Repr.repr_int` when given very large integers (#135506)

iritkatriel commented on Lib/test/test_dis.py:
```diff
@@ -1124,6 +1124,43 @@ def test_bug_46724(self):
+            ),
+        ]
+
+        for case_name, annotate_code in test_cases:
```
assert here that `annotate_code` is a code object and that `annotate_code.co_name` is `'__annotate__'`.

iritkatriel commented on Lib/test/test_dis.py:
```diff
@@ -1124,6 +1124,43 @@ def test_bug_46724(self):
+                    for instr in dis.get_instructions(annotate_code)
+                    if setup_scope_begin <= instr.offset < setup_scope_end
+                    and instr.positions
+                }
```
Why does this need to be so complicated? Don't we just need to check that all the instructions in this code object have the same line number (excluding Nones)?

AndPuQing replied:
Because the important thing is to make sure the instructions in `setup_annotations_scope `are consistent with the `end_col_offset` of RESUME, not just the lineno

no fix:

```python
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) RESUME       
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) LOAD_FAST_BORROW         # incorrect
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) LOAD_SMALL_INT           # incorrect
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) COMPARE_OP               # incorrect
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) POP_JUMP_IF_FALSE        # incorrect
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) NOT_TAKEN                # incorrect
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) LOAD_COMMON_CONSTANT     # incorrect
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) RAISE_VARARGS            # incorrect
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=2) BUILD_MAP                # incorrect
```

fix:

```python
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) RESUME
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) LOAD_FAST_BORROW
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) LOAD_SMALL_INT
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) COMPARE_OP
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) POP_JUMP_IF_FALSE
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) NOT_TAKEN
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) LOAD_COMMON_CONSTANT
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) RAISE_VARARGS
Positions(lineno=1, end_lineno=1, col_offset=0, end_col_offset=0) BUILD_MAP
```

And not all __annotate__ lines are the same, because some opcodes need to indicate the "code" where the annotation occurs.

iritkatriel replied:
Ok, so we need to check they all have the same end_col_offset as well.

iritkatriel replied:
> And not all annotate lines are the same, because some opcodes need to indicate the "code" where the annotation occurs.


The test doesn't need to work for all inputs. Only the two inputs it runs on.

AndPuQing replied:
> > And not all annotate lines are the same, because some opcodes need to indicate the "code" where the annotation occurs.
> 
> The test doesn't need to work for all inputs. Only the two inputs it runs on.

Just to be clear, instructions within code object can correspond to different line numbers. So I made an offset restriction here, limiting the check to between RESUME and BUILD_MAP (setup_annotations_scope)

iritkatriel replied:
> So I made an offset restriction here, limiting the check to between RESUME and BUILD_MAP (setup_annotations_scope)

That's not clear from the code of the test. If something changes in the future and the test needs to be updated, it will be hard to understand what the test is doing from the way it's currently written and how it needs to be fixed.  It's making a lot of assumptions about things which are current implementation details and can break at any time.

You can rewrite the locations test like is done in other cases, such as `bug1333982`/`dis_bug1333982`.
Those tests are straightforward and easy to maintain. It doesn't look at columns though.

iritkatriel replied:
Another option is to leave dis and the bytecode alone, and write a test that asserts that something which is user-visible is fixed. How did you find this problem? A traceback pointed to the wrong place?

AndPuQing replied:
> Another option is to leave dis and the bytecode alone, and write a test that asserts that something which is user-visible is fixed. How did you find this problem? A traceback pointed to the wrong place?

This issue appears to stem from an implementation choice in a downstream repository. For full context, you can refer to the issue #135700 linked in the PR.

My understanding is that the downstream project expects the position information for bytecode within `setup_annotations_scope` to have semantics similar to a `RESUME` instruction. Their reasoning seems to be that this bytecode is synthetic—it's generated by the compiler and isn't tied to any specific line of user source code, so there's no obvious "correct" location for it to point to.

Ultimately, this doesn't seem to be a factual bug, but rather a disagreement or ambiguity in the semantic interpretation of this position information.

iritkatriel replied:
Maybe @JelleZijlstra has an idea how to write a meaningful test for this.

JelleZijlstra replied:
I don't have much of an intuition on how line numbers should work for synthetic code.

iritkatriel replied:
@15r10nk are you able to suggest a meaningful unit test for this? Alternatively, can you confirm that the PR resolved the issue?

15r10nk replied:
I can verify that this fixes the issue which I have reported :+1: . I also found out that the reported bug affected debugging with pdb.

**example.py**
``` python
import example_mod
print(example_mod.__annotations__)
```

**example_mod.py**
``` python

22

class ClassVar():
    pass

__dataclass_fields__: ClassVar
``` 


```
./python -m pdb example.py 
> /home/orbit/projects/cpython/example.py(1)<module>()
-> import example_mod
(Pdb) b example_mod.py:2
Breakpoint 1 at /home/orbit/projects/cpython/example_mod.py:2
(Pdb) c
> /home/orbit/projects/cpython/example_mod.py(2)<module>()      # first hit which is expected
-> 22
(Pdb) c
> /home/orbit/projects/cpython/example_mod.py(2)__annotate__()  # pdb should not break here
-> 22
(Pdb) l
  1  	
  2 B->	22
  3  	
  4  	class ClassVar():
  5  	    pass
  6  	
  7  	__dataclass_fields__: ClassVar
  8  	
[EOF]
(Pdb) bt
  /home/orbit/projects/cpython/Lib/bdb.py(899)run()
-> exec(cmd, globals, locals)
  <string>(1)<module>()
  /home/orbit/projects/cpython/example.py(4)<module>()
-> print(example_mod.__annotations__)
> /home/orbit/projects/cpython/example_mod.py(2)__annotate__()
-> 22
(Pdb) 
```
The breakpoint is hit a second time when the synthetic code is executed which uses the incorrect source positions.

I remember that I found multiple different problems with similar synthetic codes like this in the past (`super()`,`except` related and maybe more) but I was always able to work around this issues and I did not know that they can also lead to problems when you debug with pdb. I will write a tool to find these problems in the next days.

iritkatriel replied:
Great, thank you!


The pdb example can probably be used for a unit test (in test_pdb). I'd remove the test that is currently in this PR.

15r10nk replied:
hi @iritkatriel @AndPuQing , I wrote a test with test_pdb.py maybe that can be used for this pr.

``` python
class PdbTestCase(unittest.TestCase):
    
    ...
    
    def test_issue135700(self):
        """ https://github.com/python/cpython/issues/135700 """
        module_code = """\
22

class ClassVar:
    pass
__dataclass_fields__: ClassVar
"""
        with open('testmod.py', 'w') as f:
            f.write(module_code)
        self.addCleanup(os_helper.unlink, "testmod.py")

        script = """
import testmod
print(testmod.__annotations__)
"""
        commands = """
b testmod.py:1
c
c
"""
        result = self.run_pdb_script(script, commands)
        self.assertNotIn("(1)__annotate__()", result[0])
```

sobolevn committed:
gh-135878: Fix crash in `types.SimpleNamespace.__repr__` (#135889)

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

mpage committed:
gh-135805: Document the X option and env var for controlling thread-local bytecode (#135868)

Document the X option and env var for controlling thread-local bytecode.

brianschubert committed:
gh-135855: Raise TypeError When Passing Non-dict Object to `_interpreters.set___main___attrs` (gh-135856)

brandtbucher committed:
GH-90117: Check for list and tuple before MappingView in pprint (GH-135779)

vsajip committed:
gh-91555: Revert disabling of logger while handling log record. (GH-135858)

Revert "gh-91555: disable logger while handling log record (GH-131812)"

This reverts commit 2561e148ec985755baa3984b91fd0bfc089b283c.

nascheme committed:
GH-133136: Revise QSBR to reduce excess memory held (gh-135473)

The free threading build uses QSBR to delay the freeing of dictionary
keys and list arrays when the objects are accessed by multiple threads
in order to allow concurrent reads to proceed with holding the object
lock. The requests are processed in batches to reduce execution
overhead, but for large memory blocks this can lead to excess memory
usage.

Take into account the size of the memory block when deciding when to
process QSBR requests.

Also track the amount of memory being held by QSBR for mimalloc pages.  Advance the write sequence if this memory exceeds a limit.  Advancing the sequence will allow it to be freed more quickly.

Process the held QSBR items from the "eval breaker", rather than from `_PyMem_FreeDelayed()`.  This gives a higher chance that the global read sequence has advanced enough so that items can be freed.

Co-authored-by: Sam Gross <colesbury@gmail.com>

ZeroIntensity committed:
gh-135755: Docs: C API: Document missing `PyFunction_GET*` macros (GH-135762)


Co-authored-by: Brian Schubert <brianm.schubert@gmail.com>
Co-authored-by: Petr Viktorin <encukou@gmail.com>

sobolevn committed:
gh-135839: Fix `module_traverse` and `module_clear` in `_interpchannelsmodule` (#135840)

encukou committed:
gh-135755: Move SPECIAL_ constants to a private header (GH-135922)

Macros without a `Py`/`_Py` prefix should not be defined in public headers.

encukou committed:
gh-131591: Make --without-remote-debug work (GH-135925)

The feature is checked using `defined(Py_REMOTE_DEBUG)`; defining
the macro (even as `0`) enables it.

sobolevn committed:
gh-135839: Fix `module_traverse` and `module_clear` in subinterp modules (#135937)

HarryLHW committed:
Docs: Add cross-reference for `positional_item` in the `calls` productionlist (GH-129977)

Add missing hyperlink for `positional_item`

Fidget-Spinner committed:
gh-135927: Fix MSVC Clatest C builds (GH-135935)

jtibbertsma committed:
Fix needless spinning in `_PyMutex_LockTimed` with zero timeout (gh-135872)

The free threading build could spin unnecessarily on `_Py_yield()` if the initial
compare and swap failed.

reynoldsnlp committed:
Docs: Fix indentation in `slice` class of `functions.rst` (GH-134393)

Paragraph should not be under `slice.step`. It applies to the whole class.

---------

Co-authored-by: Rob Reynolds <13379223+reynoldsnlp@users.noreply.github.com>

StanFromIreland committed:
gh-125142: remove duplicated import in `Lib/pydoc.py` (gh-135215)

nascheme committed:
Add whatsnew text for warnings module changes. (gh-135869)

terryjreedy committed:
gh-135956: Remove duplicate word in _pydatetime docstring (#135957)

_pydatetime.isoformat docstring repeats 'giving'.

encukou committed:
gh-131591: Add Py_ prefix to MAX_SCRIPT_PATH_SIZE; remove unprefixed struct tag (GH-135924)

Names/macros defined in public headers should have `Py`/`_Py` prefixes.

ZeroIntensity committed:
gh-135755: Move `PyFunction_GET_BUILTINS` to the private API (GH-135938)

encukou committed:
gh-125206: Make _Py_FFI_SUPPORT_C_COMPLEX private (GH-135932)

encukou committed:
gh-135755: Use private names (_Py*) for header file guards new in 3.14 (GH-135921)

These are private API; let's name new ones accordingly.

LamentXU123 committed:
gh-135965: Delete duplicate word in isolating-extensions howto (#135964)

Change use use to use.

dylwil3 committed:
gh-129958: New syntax error in format spec applies to both f-strings and t-strings (#135570)

Co-authored-by: Tomas R. <tomas.roun8@gmail.com>
Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

cdenihan committed:
gh-135110: Fix misleading `generator.close()` documentation (GH-135152)

The documentation incorrectly stated that generator.close() 'raises' a
GeneratorExit exception. This was misleading because the method doesn't
raise the exception to the caller - it sends the exception internally
to the generator and returns None.

duaneg committed:
gh-91555: add warning to docs about possibility of deadlock/infinite recursion (GH-135954)

* gh-91555: add warning to docs about possibility of deadlock/infinite recursion

Attempt to clarify in the documentation that care must be taken when using
multiprocessing classes to implement logging since they have builtin internal
logging, and hence may cause deadlock/infinite recursion.

* Update Doc/library/logging.handlers.rst

Co-authored-by: Vinay Sajip <vinay_sajip@yahoo.co.uk>

* Change whitespace.

---------

Co-authored-by: Vinay Sajip <vinay_sajip@yahoo.co.uk>

vstinner committed:
gh-135927: Check _MSC_VER to define _Py_NULL macro (#135987)

StanFromIreland committed:
IDLE: Update NEWS2x.txt with 2.7.0 release date (#129908)

nkorth committed:
gh-135995: Fix missing char in palmos encoding (#135990)

0x8b correctly encodes to ‹, but 0x9b was mistakenly marked as a control character instead of ›.
---------

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Brian Schubert <brianm.schubert@gmail.com>
Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

brianschubert committed:
Docs: Fix duplicate word typos (GH-135958)

freakboy3742 committed:
gh-135966: Modify iOS testbed to make app_packages a site directory (#135967)

The iOS testbed now treats the app_packages folder as a site folder. This ensures it is
on the path, but also ensures any .pth files are processed on app startup.

tim-one committed:
gh-135551: Change how sorting picks minimum run length (#135553)

New scheme from Stefan Pochmann for picking minimum run lengths.

By allowing them to change a little from one run to the next, it's possible to
arrange for that all merges, at all levels, strongly tend to be as evenly balanced
as possible, for randomly ordered data. Meaning the number of initial runs is a
power of 2, and all merges involve runs whose lengths differ by no more than 1.

freakboy3742 committed:
gh-135968: Add iOS binary stubs for strip (#135970)

Adds iOS binary stubs for invoking `strip`

serhiy-storchaka committed:
gh-53203: Fix strptime() for %c, %x and %X formats on some locales (#135971)

* Add detection of decimal non-ASCII alt digits.
* Add support of non-decimal alt digits on locale lzh_TW.
* Accept only numbers in correct range if alt digits are known.
* Fix bug in detecting the position of the week day name on locales byn_ER and wal_ET.
* Fix support of single-digit hour on locales ar_SA and bg_BG.
* Add support for %T, %R, %r, %C, %OC.
* Prepare code to use nl_langinfo().

eendebakpt committed:
gh-136017: avoid decref in rich compare for bool objects (#136018)

skirpichev committed:
gh-128051: Fix tests if sys.float_repr_style is 'legacy' (#135908)

Co-authored-by: Victor Stinner <vstinner@python.org>

serhiy-storchaka committed:
gh-78465: Fix error message for cls.__new__(cls, ...) where cls is not instantiable (GH-135981)

Previous error message suggested to use cls.__new__(), which
obviously does not work. Now the error message is the same as for
cls(...).

Fidget-Spinner committed:
gh-132732: Automatically constant evaluate pure operations (GH-132733)

This adds a "macro" to the optimizer DSL called "REPLACE_OPCODE_IF_EVALUATES_PURE", which allows automatically constant evaluating a bytecode body if certain inputs have no side effects upon evaluations (such as ints, strings, and floats).


Co-authored-by: Tomas R. <tomas.roun8@gmail.com>

hugovk committed:
gh-92266: Replace tabs with four spaces in Python files (#135983)

serhiy-storchaka committed:
gh-136028: Fix parsing month names containing "İ" (U+0130) in strptime() (GH-136029)

This affects locales az_AZ, ber_DZ, ber_MA and crh_UA.

WillChilds-Klein committed:
gh-135571: Guard `_hashlib` usage in `test_hashlib.py` (#135572)

picnixz committed:
gh-135755: rename undocumented `HACL_CAN_COMPILE_SIMD{128,256}` macros (#135847)

Rename undocumented `HACL_CAN_COMPILE_SIMD{128,256}` macros
to `_Py_HACL_CAN_COMPILE_VEC{128,256}`. These macros are private.

moreal committed:
gh-108765: fix comment about macro definitions in `_stat.c` post GH-108854 (#136027)

brandtbucher committed:
GH-135904: Optimize the JIT's assembly control flow (GH-135905)

Fidget-Spinner committed:
gh-132732: JIT: Only allow compact ints in pure evaluation (GH-136040)

LamentXU123 committed:
Fix a typo in Lib/unittest/mock.py (#136067)

Fidget-Spinner committed:
gh-132732: Fix up pure types in JIT (GH-136050)

Fix up pure types in JIT

picnixz committed:
gh-135853: add `math.signbit` (#135877)

NicolasT committed:
gh-76595: Add note on `PyCapsule_Import` behavior (GH-134022)

pablogsal committed:
gh-91048: Fix external inspection multi-threaded performance (#136005)

LamentXU123 committed:
Doc: Fix duplicate words in idlelib (#136089)

akshat62 committed:
gh-136053: Check error for TYPE_SLICE in marshal.c (GH-136054)

Fix a possible crash when deserializing a large marshal data
(at least several GiBs) containing a slice.

picnixz committed:
gh-136066: simplify `platform._platform()` (#136069)

ZeroIntensity committed:
gh-127604: Docs: Include a C stack in the `faulthandler` example (GH-136081)

fry69 committed:
gh-123299: Provide replacement for removed sqlite3 attributes in What's New 3.14 (#125566)

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

LamentXU123 committed:
Doc: fix duplicated words (#136086)

---------

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

Mr-Sunglasses committed:
gh-130160: use `.. program::` directive for documenting `venv` CLI (GH-130699)

Co-authored-by: Semyon Moroz <donbarbos@proton.me>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

sobolevn committed:
gh-136087: Remove `\r` from documented `os.linesep` values (#136088)

serhiy-storchaka committed:
gh-132813: Fix the csv documentation for quoting and escaping (#133209)

eendebakpt committed:
gh-123471: Make itertools.chain thread-safe (#135689)

xuantengh committed:
gh-135607: remove null checking of weakref list in dealloc of extension modules and objects (#135614)

Co-authored-by: Kumar Aditya <kumaraditya@python.org>
Co-authored-by: Victor Stinner <vstinner@python.org>

eendebakpt committed:
gh-123471: Make itertools.product and itertools.combinations thread-safe (#132814)

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

adamtheturtle committed:
gh-131885: Update documented signatures for `csv.{writer,reader}` (GH-136085)

StanFromIreland committed:
gh-48181: Document `codecs.charmap_build` (#135997)

vstinner committed:
gh-85702: Catch PermissionError in zoneinfo.load_tzdata() (#136117)

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

sedram committed:
gh-136122: Fix video link for math.tau documentation (#136129)

StanFromIreland committed:
gh-63207: Update `time.time` documentation after #116822 (#136068)

ericsnowcurrently committed:
gh-134939: Fill Out the concurrent.interpreters Docs (gh-135902)

sobolevn committed:
gh-135422: Fix regression in `SyntaxError` messages after #134036 (#135423)

cmaloney committed:
gh-133982: Use implementation-specific `open` in `test_fileio.OtherFileTests` (GH-135364)

encukou committed:
gh-133157: remove usage of `_Py_NO_SANITIZE_UNDEFINED` in `pyexpat` (#135346)

This was the last usage, so the macro is removed as well.

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

encukou committed:
gh-87135: threading.Lock: Raise rather than hang on Python finalization (GH-135991)

After Python finalization gets to the point where no other thread
can attach thread state, attempting to acquire a Python lock must hang.
Raise PythonFinalizationError instead of hanging.

StanFromIreland committed:
gh-133447: Update `sqlite3` What's New 3.15 entry (GH-136079)

heliang666s committed:
gh-135836: Fix `IndexError` in `asyncio.create_connection` with empty exceptions list (#135845)

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

vstinner committed:
gh-130396: Move PYOS_LOG2_STACK_MARGIN to internal headers (#135928)

Move PYOS_LOG2_STACK_MARGIN, PYOS_STACK_MARGIN,
PYOS_STACK_MARGIN_BYTES and PYOS_STACK_MARGIN_SHIFT macros to
pycore_pythonrun.h internal header. Add underscore (_) prefix to the
names to make them private. Rename _PYOS to _PyOS.

lazorikv committed:
gh-136169: Update parameter name in fractions.from_float method (#136172)

Update parameter name in fractions.from_float method

ZackerySpytz committed:
gh-87298: Add tests for find_in_strong_cache() bug in _zoneinfo (GH-24829)

Co-authored-by: Paul Ganssle <p.ganssle@gmail.com>
Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

StanFromIreland committed:
gh-105456: Remove 3 deprecated `sre_*` modules (#135994)

skirpichev committed:
gh-115119: Remove implicit fallback to the bundled libmpdec (#134078)

Co-authored-by: Victor Stinner <vstinner@python.org>
Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

Fidget-Spinner committed:
gh-136125: Use `_PyObject_GetMethodStackRef` for  `LOAD_ATTR` (GH-136127)

serhiy-storchaka committed:
gh-134280: Disable constant folding for ~ with a boolean argument (GH-134982)

This moves the deprecation warning from compile time to run time.

colesbury committed:
gh-134009: Expose `PyMutex_IsLocked` in the public C API (gh-134365)

The `PyMutex_IsLocked()` function is useful in assertions for verifying
that code maintains certain locking invariants.

Fidget-Spinner committed:
gh-136183: Deal with escapes in JIT optimizer's constant evaluator (GH-136184)

skirpichev committed:
gh-115119: Defer removal of bundled libmpdec to 3.16 (#133997)

Rename libmpdecimal -> libmpdec

see https://www.bytereef.org/mpdecimal/doc/libmpdec/index.html

garry-cairns committed:
gh-134567: Add the formatter parameter in unittest.TestCase.assertLogs (GH-134570)

sobolevn committed:
gh-136193: Improve `TypeError` msg when comparing two `SimpleNamespace`s (#136195)

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

vstinner committed:
gh-127705: Move Py_INCREF_MORTAL() to the internal C API (GH-136178)

Rename Py_INCREF_MORTAL() to _Py_INCREF_MORTAL() and move it to
pycore_object.h internal header.

sobolevn committed:
gh-136203: Improve `TypeError` msg when comparing two `MappingProxyType`s (#136204)

Co-authored-by: Victor Stinner <vstinner@python.org>

sobolevn committed:
Replace `capi-sig` mailing list with `discuss.python.org` (#136211)

LamentXU123 committed:
gh-136135: Doc: Fix some broken links (GH-136137)

serhiy-storchaka committed:
gh-135836: Fix `IndexError` in `asyncio.create_connection()` (#135875)

abstractedfox committed:
gh-135640: Adds more type checking to ElementTree (GH-135643)

yoney committed:
Fix comments for `heapq.siftup_max` (#135359)

Co-authored-by: mpage <mpage@meta.com>

danielhollas committed:
gh-135069: Fix exception message in encodings.idna module (#135071)

serhiy-storchaka committed:
gh-133740: Fix regression in locale.nl_langinfo(ALT_DIGITS) (GH-136237)

There is no need to temporary switch locale for items ALT_DIGITS and ERA
if the nl_langinfo() result is empty (most locales).

skirpichev committed:
gh-130664: Treat '0' fill character with align '=' as zero-padding for Fraction's (GH-131067)

Eclips4 committed:
`Python/gc.c`: Refer to `InternalDocs` instead of devguide. (#136243)

skirpichev committed:
gh-115119: Recommend upstream libmpdec in build requirements (#136205)

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

vstinner committed:
gh-135075: Make PyObject_SetAttr() fail with NULL value and exception (#136180)

Make PyObject_SetAttr() and PyObject_SetAttrString() fail if called
with NULL value and an exception set.

emmatyping committed:
gh-135252: Document Zstandard integration across zipfile, shutil, and tarfile (#135311)

Document Zstandard integration across zipfile, shutil, and tarfile

serhiy-storchaka committed:
gh-135661: Fix parsing start and end tags in HTMLParser according to the HTML5 standard (GH-135930)

* Whitespaces no longer accepted between `</` and the tag name.
  E.g. `</ script>` does not end the script section.

* Vertical tabulation (`\v`) and non-ASCII whitespaces no longer recognized
  as whitespaces. The only whitespaces are `\t\n\r\f `.

* Null character (U+0000) no longer ends the tag name.

* Attributes and slashes after the tag name in end tags are now ignored,
  instead of terminating after the first `>` in quoted attribute value.
  E.g. `</script/foo=">"/>`.

* Multiple slashes and whitespaces between the last attribute and closing `>`
  are now ignored in both start and end tags. E.g. `<a foo=bar/ //>`.

* Multiple `=` between attribute name and value are no longer collapsed.
  E.g. `<a foo==bar>` produces attribute "foo" with value "=bar".

* Whitespaces between the `=` separator and attribute name or value are no
  longer ignored. E.g. `<a foo =bar>` produces two attributes "foo" and
  "=bar", both with value None; `<a foo= bar>` produces two attributes:
  "foo" with value "" and "bar" with value None.

* Fix Sphinx errors.

* Apply suggestions from code review

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

* Address review comments.

* Move to Security.

---------

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

brettcannon committed:
Clarify some details regarding `sys.monitoring` (#133981)

cmaloney committed:
gh-133982: Test _pyio.BytesIO in free-threaded tests (gh-136218)

kumaraditya303 committed:
gh-115999: remove redundant check in free-threading from `_STORE_ATTR_WITH_HINT` (#136249)

kumaraditya303 committed:
gh-129824: fix data races in subinterpreters under TSAN (#135794)

This fixes the data races in typeobject.c in subinterpreters under free-threading. The type flags and slots are only modified in the main interpreter as all static types are first initialised in main interpreter.

serhiy-storchaka committed:
gh-102555: Fix comment parsing in HTMLParser according to the HTML5 standard (GH-135664)

* "--!>" now ends the comment.
* "-- >" no longer ends the comment.
* Support abnormally ended empty comments "<-->" and "<--->".

---------

Co-author: Kerim Kabirov <the.privat33r+gh@pm.me>
Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

rickeylev committed:
gh-135773: have pyvenv.cfg without home key anchor a venv and deduce home (#135831)

This is still formally undefined behaviour, but we may as well
keep the *same* undefined behaviour as previous versions.

PEP 796 proposes a cleaner and more consistent replacement for 3.15+

rffontenelle committed:
Docs: Move "or" outside monospace syntax in `tarfile.rst` (GH-136263)

wsfulton committed:
gh-136288: Fix error message in `_testcapi/vectorcall.c` (GH-136258)

Use the %N format specifier instead of %s and `PyType_GetName`.

sobolevn committed:
gh-136297: Test all `pickle` protocols in `test_zoneinfo_property.py` (#136298)

LamentXU123 committed:
gh-101100: Fix references in `http.cookiejar` docs (GH-136238)

Viicos committed:
gh-130870: Preserve `GenericAlias` subclasses in `typing.get_type_hints()` (#131583)

youknowone committed:
gh-136047: Allow typing._allow_reckless_class_checks to check `_py_abc` (#136115)

emmatyping committed:
gh-136315: Fix skipped multithreading test in test_zstd (#136320)

Fix skipped test in test_zstd

mattwang44 committed:
gh-136032: Fix `argparse.BooleanOptionalAction` doc (#136133)

sobolevn committed:
gh-136285: Improve `pickle` protocol testing in `test_interpreters` (#136286)

kumaraditya303 committed:
gh-109700: fix memory error handling in `PyDict_SetDefault` (#136338)

aisk committed:
gh-109070: Document that get_context in multiprocessing have side effect (#136341)

Document that get_context in multiprocessing have side effect

sobolevn committed:
gh-136021: Make `type_params` a required parameter for `typing._eval_type` (#136332)

JelleZijlstra committed:
gh-136316: Make typing.evaluate_forward_ref better at evaluating nested forwardrefs (#136319)

serhiy-storchaka committed:
gh-136289: Fix test_sqlite3 on platforms with strict UTF-8 filesystem (GH-136326)

kumaraditya303 committed:
gh-109700: fix interpreter finalization while handling memory error (#136342)

skirpichev committed:
gh-87790: support thousands separators for formatting fractional part of Decimal (#132202)

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

skirpichev committed:
gh-87790: support thousands separators for formatting fractional part of Fraction (#132204)

vsajip committed:
gh-94503: Update logging cookbook with an example of uniformly handling newlines in output. (GH-136217)

vstinner committed:
gh-127502: Remove XML vulnerability table (GH-135294)

* Remove the table
* Replace warnings with notes


Latest releases of Python 3.9-3.15 include expat 2.7.1 which is not vulnerable.

expat 2.6.0 was released in February 2024.

encukou committed:
gh-135755: Document __future__.* and CO_* as proper Sphinx objects (GH-135980)

* Turn the __future__ table to list-table.
  This'll make it easier to add entries that need longer markup
* Semantic markup for __future__ feature descriptions.
* Document CO_* C macros.

sobolevn committed:
gh-101100: Fix sphinx warnings in `whatsnew/3.9` (#136163)

encukou committed:
gh-131591: Document Py_REMOTE_DEBUG (GH-135929)

asottile committed:
gh-86682: Add versionadded for sys._getframemodulename (#136325)

add versionadded for sys._getframemodulename

encukou committed:
gh-135913: Document ob_refcnt, ob_type, ob_size (GH-135914)

* gh-135913: Document ob_refcnt, ob_type, ob_size

In `typeobj.rst`, instead of `:c:member:` it would be better to
use `.. c:member::` with a `:no-index:` option, see:

See ref. https://www.sphinx-doc.org/en/master/usage/domains/index.html#basic-markup

However, `c:member` currently does not support `:no-index:`.

serhiy-storchaka committed:
gh-124486: Fix test_whichdb_ndbm in test_dbm on NetBSD (GH-136335)

On NetBSD, ndbm.open() does not fail for empty file.

m-aciek committed:
gh-136155: Docs: only add custom OpenGraph protocol meta tags for HTML builds (#136187)

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>

encukou committed:
gh-120713: Make _Py_NORMALIZE_CENTURY private (GH-135933)

LamentXU123 committed:
gh-101100: Fix Sphinx warnings in library/email.compat32-message.rst (#136323)

clin1234 committed:
gh-135906: Use `_PyObject_CAST` in internal headers (GH-135892)

Fixes build errors encountered in python-greenlet/greenlet#450 when building greenlet on the free-threaded build.

---------

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

kumaraditya303 committed:
gh-98388: add tests for happy eyeballs  (#136368)

hugovk committed:
gh-102567: Add missing newline to `--help-all` (GH-136391)

LamentXU123 committed:
gh-101100: Fix sphinx warnings in `whatsnew/3.11.rst` (#136402)

kumaraditya303 committed:
gh-134043: use `_PyObject_GetMethodStackRef` in pattern matching (#136356)

kumaraditya303 committed:
gh-117657: enable `test_capi` under TSAN (#136269)

sobolevn committed:
gh-136297: Fix `hypothesis` and `subTest` usage in `test_zoneinfo_property.py` (#136384)

Commit:
gh-101100: Fix sphinx warnings in `Doc/library/exceptions.rst` (#136309)

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

pablogsal committed:
gh-136186: Fix race condition in test_external_inspection.test_only_active_thread (#136347)

aisk committed:
gh-136380: Fix import behavior for `concurrent.futures.InterpreterPoolExecutor` (#136381)

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>
Co-authored-by: Peter Bierma <zintensitydev@gmail.com>
Co-authored-by: sobolevn <mail@sobolevn.me>

vstinner committed:
gh-136156: Allow using linkat() with TemporaryFile (#136281)

tempfile.TemporaryFile() no longer uses os.O_EXCL with os.O_TMPFILE,
so it's possible to use linkat() on the file descriptor.

hugovk committed:
Update bytecode magic number in tests for the 3.14 release candidate (#136427)

nascheme committed:
GH-91636: Clear weakrefs created by finalizers. (GH-136401)

Weakrefs to unreachable garbage that are created during running of
finalizers need to be cleared.  This avoids exposing objects that
have `tp_clear` called on them to Python-level code.

StanFromIreland committed:
gh-136162: Document `encodings` package functions (#136164)

Closes #136162.

senyai committed:
gh-92536: Fix comment about number of unicode string types (#136439)

pablogsal committed:
gh-91048: Revert the memory cache removal for remote debugging (#136440)

gh-91048: Reintroduce the memory cache for remote debugging

hoodmane committed:
gh-136229: Remove Platform Emscripten is not supported warning (#136230)

Updates configure script to identify Emscripten as Tier 3.

dg-pb committed:
gh-119109: improve `functools.partial` vectorcall with keywords (#124584)

Co-authored-by: Kumar Aditya <kumaraditya@python.org>
Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

vsajip committed:
gh-94503: Update logging cookbook example with info on addressing log injection. (GH-136446)

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

StanFromIreland committed:
gh-53243: Document `codecs.readbuffer_encode()` (#136284)

Closes #53243

JelleZijlstra committed:
gh-134657: Remove newly added private names from asyncio.__all__ (#134665)

injust committed:
gh-136447: Use `self.loop` instead of global `loop` variable in asyncio REPL (#136448)

geofft committed:
Docs: unittest.enterModuleContext is not a classmethod (#136464)

NekrodNIK committed:
gh-131825: Fix `sqlite3` timezone-naive adapter recipe (GH-136270)

bionade24 committed:
gh-81520: Document unexpected `os.path.ismount` behaviour with btrfs subvolumes (GH-136058)

rhettinger committed:
Minor edit: Move comments closer to the code they describe (gh-136477)

brandtbucher committed:
GH-115802: Optimize JIT stencils for size (GH-136393)

diegorusso committed:
Add Diego Russo as code owner of the JIT (#136460)

zware committed:
gh-136145: Define 'standard library' and 'stdlib' in the glossary (GH-136146)

---------

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>
Co-authored-by: Éric <merwok@netwok.org>

vstinner committed:
gh-136156: Skip test_tempfile.test_link_tmpfile() on Android (#136430)

Adds a test skip on platforms where hard links are not available (which includes Android).

vstinner committed:
gh-102740: Clarify time.monotonic() "system-wide" in the doc (#136431)

pablogsal committed:
gh-136476: Show the full stack in get_async_stack_trace in _remote_debugging (#136483)

emmatyping committed:
gh-135846: Add zstd dependency to Android build script (#136253)

Adds zstd to the Android build process.

---------

Co-authored-by: Malcolm Smith <smith@chaquo.com>

encukou committed:
gh-136476: Remove creation of unused list (GH-136494)

sobolevn committed:
gh-136438: Make sure `test_builtins` pass with all optimization levels (#136474)

encukou committed:
gh-136209: Add .. c:var:: declarations for C exception types (GH-136210)


Co-authored-by: Victor Stinner <vstinner@python.org>

eendebakpt committed:
gh-82088: Improve performance of PyLong_As*() for multi-digit ints (#135585)

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

Rogdham committed:
gh-136394: Fix race condition in test_zstd (GH-136432)

davepeck committed:
gh-132661: Add default value (of `""`) for `Interpolation.expression` (#136441)

StanFromIreland committed:
gh-82663: Clarify `codecs.iterdecode/encode` docs (#136497)

Closes #82663

StanFromIreland committed:
gh-52876: Implement missing parameter in `codecs.StreamReaderWriter` functions (#136498)

Closes #52876

lkollar committed:
gh-135953:  Implement sampling tool under profile.sample (#135998)

Implement a statistical sampling profiler that can profile external
Python processes by PID. Uses the _remote_debugging module and converts
the results to pstats-compatible format for analysis.


Co-authored-by: Pablo Galindo <pablogsal@gmail.com>

sergey-miryanov committed:
gh-136517: Print uncollectable objects if DEBUG_UNCOLLECTABLE mode was set (#136518)

brandtbucher committed:
Omit `Python/perf_jit_trampoline.c` from the `**/*jit*` CODEOWNERS rule (#136519)

Omit perf_jit_trampoline from "JIT" codeowners

ethanfurman committed:
gh-107538: [Enum] fix handling of inverted/negative values (GH-132273)

* Fix flag mask inversion when unnamed flags exist.

For example:

    class Flag(enum.Flag):
        A = 0x01
        B = 0x02
        MASK = 0xff

    ~Flag.MASK is Flag(0)

* EJECT and KEEP flags (IntEnum is KEEP) use direct value.

* correct Flag inversion to only flip flag bits

IntFlag will flip all bits -- this only makes a difference in flag sets with
missing values.

* correct negative assigned values in flags

negative values are no longer used as-is, but become inverted; i.e.

    class Y(self.enum_type):
        A = auto()
        B = auto()
        C = ~A        # aka ~1 aka 0b1 110 (from enum.bin()) aka 6
        D = auto()

    assert Y.C. is Y.B|Y.D

LamentXU123 committed:
gh-101100: Fix sphinx warnings in `library/email.parser.rst` (#136475)

StanFromIreland committed:
gh-76637: Note that `undefined` Codec is for testing (#136531)

Closes #76637

AndPuQing committed:
add test_pdb case
